// truffle-decoder/dist/types/pointer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function isStackPointer(pointer) {
    return typeof pointer !== "undefined" && "stack" in pointer;
}
exports.isStackPointer = isStackPointer;
function isMemoryPointer(pointer) {
    return typeof pointer !== "undefined" && "memory" in pointer;
}
exports.isMemoryPointer = isMemoryPointer;
function isCalldataPointer(pointer) {
    return typeof pointer !== "undefined" && "calldata" in pointer;
}
exports.isCalldataPointer = isCalldataPointer;
function isStoragePointer(pointer) {
    return typeof pointer !== "undefined" && "storage" in pointer;
}
exports.isStoragePointer = isStoragePointer;
function isStackLiteralPointer(pointer) {
    return typeof pointer !== "undefined" && "literal" in pointer;
}
exports.isStackLiteralPointer = isStackLiteralPointer;
function isConstantDefinitionPointer(pointer) {
    return typeof pointer !== "undefined" && "definition" in pointer;
}
exports.isConstantDefinitionPointer = isConstantDefinitionPointer;
